File: 单调栈/3. 下一个更大元素 II - 503/Next Greater Element II.py
# 23/03/27
# https://leetcode.cn/problems/next-greater-element-ii/

from collections import List 

def nextGreaterElements(nums: List[int]) -> List[int]:

    lengthOfNums = len(nums)
    stack = list()
    result = [-1 for _ in range(lengthOfNums)]

    for i in range(lengthOfNums * 2):
        index = i % lengthOfNums
        while stack and nums[index] > nums[stack[-1]]:
            result[stack[-1]] = nums[index]
            stack.pop()
        stack.append(index)
    
    return result

File: 单调栈/3. 下一个更大元素 II - 503/Next Greater Element II.cpp
/*
    23/03/27
    https://leetcode.cn/problems/next-greater-element-ii/
 */

#include <iostream>
#include <vector>
#include <stack>

using namespace std;

/*
    solution: 栈
    result: ✅
    TC: O(n)
    SC: O(1)
 */

vector<int> nextGreaterElements(vector<int>& nums) {

    int lengthOfNums = nums.size();
    vector<int> result(lengthOfNums, -1);
    stack<int> st;

    for (int i = 0; i < lengthOfNums * 2; i ++) 
    {
        int index = i % lengthOfNums;
        while (!st.empty() && nums[index] > nums[st.top()])
        {
            result[st.top()] = nums[index];
            st.pop();
        }
        st.push(index);
    }

    return result;
}